pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -263,36 +304,144 @@ func buildHTTPVirtualServices(obj config.Config, gateways map[parentKey]map[k8s.
 			}}
 		}
 
-		route, err := buildHTTPDestination(r.BackendRefs, obj.Namespace, domain, zero)
+		route, err := buildHTTPDestination(r.BackendRefs, ns, domain, zero)
 		if err != nil {
 			reportError(err)
-			return nil
+			return
 		}
 		vs.Route = route
 
 		httproutes = append(httproutes, vs)
 	}
 	reportError(nil)
+
 	gatewayNames := referencesToInternalNames(parentRefs)
 	if len(gatewayNames) == 0 {
-		return nil
+		return
+	}
+	count := 0
+	for _, gw := range gatewayNames {
+		if gw == "mesh" {
+			// it's a mesh route
+			if _, f := meshRoutes[ns]; !f {
+				meshRoutes[ns] = make(map[string]*config.Config)
+			}
+			for _, host := range hosts {
+				if cfg := meshRoutes[ns][host]; cfg != nil {
+					// merge http routes
+					vs := cfg.Spec.(*istio.VirtualService)
+					vs.Http = append(vs.Http, httproutes...)
+					sortHTTPRoutes(vs.Http)
+				} else {
+					sortHTTPRoutes(httproutes)
+					// build one VS per namespace+host
+					meshRoutes[ns][host] = &config.Config{
+						Meta: config.Meta{
+							CreationTimestamp: obj.CreationTimestamp,
+							GroupVersionKind:  gvk.VirtualService,
+							Name:              fmt.Sprintf("%s-%d", name, count),
+							Annotations:       parentMeta(obj, nil),
+							Namespace:         ns,
+							Domain:            domain,
+						},
+						Spec: &istio.VirtualService{
+							Hosts:    []string{host},
+							Gateways: []string{gw},
+							Http:     httproutes,
+						},
+					}
+					count++
+				}
+			}
+			continue
+		}
+		// it's a gateway route
+		if _, f := gatewayRoutes[gw]; !f {
+			gatewayRoutes[gw] = make(map[string]*config.Config)
+		}
+		for _, host := range hosts {
+			if cfg := gatewayRoutes[gw][host]; cfg != nil {
+				// merge http routes
+				vs := cfg.Spec.(*istio.VirtualService)
+				vs.Http = append(vs.Http, httproutes...)
+				sortHTTPRoutes(vs.Http)
```

Comment:
sortHTTPRoutes is n^2 because we call it each time we append. Maybe we can just call it once on all routes after?